mne/decoding/search_light.py

```diff
@@ -0,0 +1,332 @@
+            Transformed values generated by each estimator.
+        """
+        method = 'transform'
+        if not hasattr(self.base_estimator, 'transform'):
+            method = 'predict'
+        self._transform(X, method)
+
+    def predict(self, X):
+        """Predict the dataset with a series of independent estimators.
+
+        Parameters
+        ----------
+        X : array, shape(n_samples, n_features, n_iteration)
+            The training input samples. For each iteration, a clone estimator
+            is fitted independently.
+
+        Returns
+        -------
+        y_pred : array, shape(n_samples, n_train_iterations, n_test_iterations)
+            Predicted values for each estimator.
+        """
+        return self._transform(X, 'predict')
+
+    def predict_proba(self, X):
+        if not hasattr(self.base_estimator, 'predict_proba'):
+            ValueError('self.base_estimator does not have `predict_proba` '
+                       'method.')
+        return self._transform(X, 'predict_proba')
+
+    def decision_function(self, X):
+        if not hasattr(self.base_estimator, 'decision_function'):
+            ValueError('self.base_estimator does not have `decision_function` '
+                       'method.')
+        return self._transform(X, 'decision_function')
+
+    def score(self, X, y):
+        # FIXME: not sure how to design
+        raise NotImplementedError()
+
+
+def _gl_transform(estimators, X, method):
+    """Transform the dataset by applying each estimator to all iteration of
+    the data.
+
+    Parameters
+    ----------
+    X : array, shape(n_samples, n_features, n_iteration)
+        The training input samples. For each iteration, a clone estimator
+        is fitted independently.
+
+    Returns
+    -------
+    Xt : array, shape(n_samples, n_iterations)
+        Transformed values generated by each estimator.
+    """
+    n_sample, n_chan, n_iter = X.shape
+    for ii, est in enumerate(estimators):
+        # stack generalized data for faster prediction
+        X_stack = np.transpose(X, [1, 0, 2])
+        X_stack = np.reshape(X_stack, [n_chan, n_sample * n_iter]).T
+        transform = getattr(est, method)
+        _y_pred = transform(X_stack)
+        # unstack generalizations
+        if _y_pred.ndim == 2:
+            _y_pred = np.reshape(_y_pred, [n_sample, n_iter])
+        else:
+            shape = np.r_[n_sample, n_iter, _y_pred.shape[1:]]
+            _y_pred = np.reshape(_y_pred, shape)
+        # init
+        if ii == 0:
+            y_pred = _gl_init_pred(_y_pred, X, len(estimators))
+        y_pred[:, ii, ...] = _y_pred
+    return y_pred
+
+
+def _gl_init_pred(y_pred, X, n_train):
+    n_sample, n_chan, n_iter = X.shape
+    if y_pred.ndim == 3:
+        y_pred = np.zeros((n_sample, n_train, n_iter, y_pred.shape[-1]))
+    else:
+        y_pred = np.zeros((n_sample, n_train, n_iter))
+    return y_pred
```

Comment:
Jona did you take a picture by chance?

On Sat, Jul 2, 2016 at 8:00 PM jona-sassenhagen <notifications@github.com>
wrote:

> In mne/decoding/search_light.py
> <https://github.com/mne-tools/mne-python/pull/3381#discussion_r69381020>:
>
> > +            shape = np.r_[n_sample, n_iter, _y_pred.shape[1:]]
> > +            _y_pred = np.reshape(_y_pred, shape)
> > +        # init
> > +        if ii == 0:
> > +            y_pred = _gl_init_pred(_y_pred, X, len(estimators))
> > +        y_pred[:, ii, ...] = _y_pred
> > +    return y_pred
> > +
> > +
> > +def _gl_init_pred(y_pred, X, n_train):
> > +    n_sample, n_chan, n_iter = X.shape
> > +    if y_pred.ndim == 3:
> > +        y_pred = np.zeros((n_sample, n_train, n_iter, y_pred.shape[-1]))
> > +    else:
> > +        y_pred = np.zeros((n_sample, n_train, n_iter))
> > +    return y_pred
>
> Btw did you see the poster on the inapplicability of tests against chance
> when there is by definition no below chance (e.g. in not-cross decoding)?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/3381/files/02b9b7688bd19d76853acd3864153dfe3eccdfb7#r69381020>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AB0filKViCAA7WryejM9fDpZoajAnH0Rks5qRqc9gaJpZM4JDsU1>
> .
>